feat: Support property based authorization checks (engine)

ce-dmelnych opened the pull request:
## Description

This PR introduces property-based authorization checks in the Zeebe engine, enabling more flexible and dynamic authorization policies that grant permissions based on runtime resource properties rather than static resource IDs.

### Motivation

Resource ID-based authorization requires explicitly granting permissions for specific resource identifiers (e.g., process definition IDs) or to all resources (WILDCARD `*`). However, for user tasks, this approach is limiting because:
1. **Task assignments are dynamic** — Users need to operate on tasks they're assigned to or are candidates for, but the specific task IDs aren't known in advance. Organizations can to grant permissions like "users can complete tasks they are assigned to" without enumerating every possible task key.
2. **Reduced administrative overhead** — Property-based rules eliminate the need to create individual permission entries for each resource.

Property-based authorization solves this by allowing permissions to be granted based on resource attributes (e.g., `assignee`, `candidateUsers`, `candidateGroups`) that are evaluated at runtime.

## Main highlights

### New capabilities

- Property-based authorization for `USER_TASK` resources
  - Currently, property-based authorization is only supported for the `USER_TASK` authorization resource type
- Supported properties:
  - `assignee` — Matches when the requesting user is the task assignee
  - `cancidateUsers` — Matches when the requesting user is in the candidate users list
  - `candidateGroups` — Matches when any of the requesting principal's groups overlap with candidate groups
  
### Key changes

1. `AuthorizationRequest` refactoring:
   - Replaced private constructor with builder-based construction for improved readability
   - Added explicit validation for required fields and mutually exclusive inputs
   - Extended with `resourceProperties` support alongside existing `resourceIds`
   - Moved error message constants locally for better cohesion

2. New authorization evaluation pipeline:
   - Added `PropertyAuthorizationEvaluatorRegistry` for registering property evaluators per resource type
   - Introduced `PropertyAuthorizationEvaluator` interface for resource-type-specific property matching
   - Implemented `UserTaskPropertyAuthorizationEvaluator` for user task property evaluation
   - Authorization checks now follow a 3-step process:  primary entity -> mapping rules -> property-based (if properties present in request).

3. Group resolution for property matching:
   - Groups are collected from all sources:  user/client membership and matching mapping rules

## Adoption guide

To add property-based authorization support for other `AuthorizationResourceType` values:

### 1. Create a properties type

Implement the `ResourceAuthorizationProperties` interface for your resource type:

<details>
<summary><code>YourResourceAuthorizationProperties.java</code></summary>

```java
public record YourResourceAuthorizationProperties(
    String property1,
    List<String> property2
) implements ResourceAuthorizationProperties {

  public static final String PROP_PROPERTY1 = "property1";
  public static final String PROP_PROPERTY2 = "property2";

  @Override
  public boolean hasProperties() {
    return hasProperty1() || hasProperty2();
  }

  @Override
  public Set<String> getPropertyNames() {
    final Set<String> names = new HashSet<>();
    if (hasProperty1()) names.add(PROP_PROPERTY1);
    if (hasProperty2()) names.add(PROP_PROPERTY2);
    return names;
  }

  public boolean hasProperty1() {
    return property1 != null && ! property1.isEmpty();
  }

  public boolean hasProperty2() {
    return property2 != null && !property2.isEmpty();
  }

  public static Builder builder() {
    return new Builder();
  }

  // ... [Optional] Builder class ... 
}
```
</details>

### 2. Create a property evaluator

Implement the `PropertyAuthorizationEvaluator` interface:
<details>
<summary><code> YourPropertyAuthorizationEvaluator.java</code></summary>

```java
public final class YourPropertyAuthorizationEvaluator
    implements PropertyAuthorizationEvaluator<YourResourceAuthorizationProperties> {

  @Override
  public AuthorizationResourceType resourceType() {
    return AuthorizationResourceType.YOUR_RESOURCE_TYPE;
  }

  @Override
  public Class<YourResourceAuthorizationProperties> propertiesType() {
    return YourResourceAuthorizationProperties.class;
  }

  @Override
  public Set<String> evaluateMatchingProperties(
      final Map<String, Object> claims,
      final YourResourceAuthorizationProperties properties) {
    // Determine which properties the principal matches
    // Return the set of matched property names
  }
}
```
</details>

### 3. Register evaluator

Add your evaluator to the registry in `AuthorizationCheckBehavior` constructor:

```java
propertyEvaluatorRegistry =
    new PropertyAuthorizationEvaluatorRegistry()
        .register(new UserTaskPropertyAuthorizationEvaluator(claimsExtractor, mappingRuleState))
        .register(new YourPropertyAuthorizationEvaluator(/* dependencies */));
```

### 4. Use in authorization requests

```java
AuthorizationRequest.builder()
    .command(command)
    .resourceType(AuthorizationResourceType.YOUR_RESOURCE_TYPE)
    .permissionType(PermissionType.UPDATE)
    .resourceProperties(
        YourResourceAuthorizationProperties.builder()
            .property1(resource.getProperty1())
            .property2(resource.getProperty2())
            .build())
    .build();
```

## Related issues

closes #40114

ce-dmelnych committed:
refactor: simplify AuthorizationRequest construction via builder

Replace the private constructor with many parameters by a builder-based constructor
to improve readability and maintainability.

Even though the constructor was private, the growing parameter list made it hard
to reason about parameter ordering, semantics, and future extensions. Passing the
Builder directly avoids this issue and keeps the object creation logic cohesive.

Additionally, extract small private helper methods to resolve derived fields,
keeping the constructor declarative and easy to read.

ce-dmelnych committed:
refactor: validate AuthorizationRequest builder invariants

Add explicit validation to `AuthorizationRequest.Builder#build()` to enforce required
fields and mutually exclusive authorization inputs.

This acts as a safeguard against invalid `AuthorizationRequest` setups by ensuring:
- `resourceType` and `permissionType` are always provided
- exactly one authorization source is used (`command` or `authorizationClaims`)

Fail-fast validation makes incorrect usage explicit and prevents constructing
inconsistent authorization requests.

Change `TenantResolver#isAssignedToTenant` to accept `claims` and `tenantId` directly,
avoiding unnecessary dependency on `AuthorizationRequest` when only these inputs are required.

ce-dmelnych committed:
refactor: move AuthorizationRequest-specific error messages locally

Move error message constants used exclusively by `AuthorizationRequest` from
`AuthorizationCheckBehavior` into `AuthorizationRequest` itself.

This reduces unnecessary coupling between the classes and keeps error message
definitions close to their actual usage, improving cohesion and readability.

ce-dmelnych committed:
feat: allow AuthorizationRequest to carry resource properties

Extend AuthorizationRequest with optional `resourceProperties` that can be provided
instead of `resourceIds` when constructing an authorization request.

Mutual exclusivity between `resourceIds` and `resourceProperties` is enforced
to ensure the request represents a single authorization scope, either ID- or
property-based.

ce-dmelnych committed:
test: add coverage for AuthorizationRequest validation and error messages

Add unit tests for `AuthorizationRequest` to verify builder validation rules,
resource identifier and property handling, and generated authorization error
messages.

ce-dmelnych committed:
refactor: add AUTHORIZED constant for authorization check results

Add a private static constant `AUTHORIZED = Either.right(null)` to
AuthorizationCheckBehavior for improved code readability and consistency
when returning successful authorization results.

ce-dmelnych committed:
feat: support property-based authorization checks for user tasks

Introduce support for property-based authorization checks that allow
granting permissions based on user task properties ("assignee",
"candidateUsers", "candidateGroups") rather than specific resource IDs.

This enables flexible authorization policies where users can be granted
permissions to operate on tasks they are assigned to or are candidates for.

npepinpe reviewed (commented):
❌ I'm not quite done, but already it strikes me that the generic resource property map is the wrong approach. Why not leverage the type system? This is an approach I see happen more commonly across the code base, and I'm not super keen on it. I also can't quite explain why people choose it yet, but that's a bit besides the point here.

If we use a generic map, it means we always have to allocate a new map, we always have to check the type of the property and cast it - it also begs the question, what if the type is different? What happens? What _should_ be the semantics?

Why not just parameterize the evaluator interface and enforce the mapping between resource and evaluator? I get that this approach tends to infect adjacent types (e.g. `AuthorizationRequest`), so I get that it's not ideal. But still, I'm just not sold on all these generic maps everywhere, with string constant keys littering our code base 🤔 

As there's no example of usage here that I could find (other than tests), maybe I just need some to better understand how that's going to work.

But I would prefer we look into something that is _typed_ and not just maps of strings.

ce-dmelnych commented:
> ❌ I'm not quite done, but already it strikes me that the generic resource property map is the wrong approach. Why not leverage the type system? This is an approach I see happen more commonly across the code base, and I'm not super keen on it. I also can't quite explain why people choose it yet, but that's a bit besides the point here.
> 
> If we use a generic map, it means we always have to allocate a new map, we always have to check the type of the property and cast it - it also begs the question, what if the type is different? What happens? What _should_ be the semantics?
> 
> Why not just parameterize the evaluator interface and enforce the mapping between resource and evaluator? I get that this approach tends to infect adjacent types (e.g. `AuthorizationRequest`), so I get that it's not ideal. But still, I'm just not sold on all these generic maps everywhere, with string constant keys littering our code base 🤔
> 
> As there's no example of usage here that I could find (other than tests), maybe I just need some to better understand how that's going to work.
> 
> But I would prefer we look into something that is _typed_ and not just maps of strings.

---

Hi @npepinpe, thank you for the feedback!  👋

You raise a valid point. I initially went with the `Map<String, Object>` approach because it was straightforward for handling a variety of property types, and mirrored the same map pattern used for representing command authorizations (claims).

However, I agree that leveraging the type system is a better approach here 👍🏻

I'll work on refactoring this to use typed properties and will ping you once it's ready for another look!

Thanks again for pushing back on this 🙏

ce-dmelnych committed:
refactor: replace Map<String, Object> with typed resource properties

Replace the generic Map<String, Object> approach for resource properties with
a type-safe model using interfaces and records.  This addresses code review
feedback about leveraging the type system instead of runtime type checking.

Ben-Sheppard reviewed (commented):
There are quite a few changes in this PR so it feels daunting/challenging to review effectively however there a re large parts which are boilerplate so it looks larger than it is.

I have some questions/comments around efficiency/naming but some are NABD.

ce-dmelnych committed:
refactor: rename matches to evaluateMatchingProperties

Rename `PropertyAuthorizationEvaluator.matches(..)` to `evaluateMatchingProperties(..)`
to better reflect that the method returns matched property names rather than
a boolean result.

ce-dmelnych commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/authorization/property/UserTaskAuthorizationProperties.java:
```diff
@@ -0,0 +1,101 @@
+  public static final String PROP_CANDIDATE_GROUPS = "candidateGroups";
+
+  public UserTaskAuthorizationProperties {
+    candidateUsers = candidateUsers != null ? List.copyOf(candidateUsers) : null;
```
Yes, empty means the task has no candidate users/groups, so those properties won't grant access, nothing to match against.

I'll also adjust this constructor to default to empty collections instead of `null` to avoid nullability checks for consumers.

ce-dmelnych committed:
refactor: rename checkPropertyBasedAuthorization to isAuthorizedByProperties

ce-dmelnych commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/authorization/AuthorizationCheckBehavior.java:
```diff
@@ -271,6 +283,67 @@ private AuthorizationResult checkMappingRuleAuthorization(
     return new AuthorizationResult(tenantAccess, resourceAccess);
   }
 
+  private boolean checkPropertyBasedAuthorization(
```
Good point!
Renamed to `isAuthorizedByProperties` ✅

ce-dmelnych reviewed (commented):
@Ben-Sheppard, @npepinpe 
All comments addressed.
Ready for another round of review 🙇🏻

Ben-Sheppard commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/authorization/AuthorizationCheckBehavior.java:
```diff
@@ -180,33 +184,40 @@ public Either<Rejection, Void> isAnyAuthorizedOrInternalCommand(
+
+    // Step 3: Check if request can be authorized by resource properties
+    // Only if we have tenant access (from primary or mapping rules)
+    if (request.hasResourceProperties() && mappingRuleResult.hasTenantAccess()) {
```
❓ The comment here mentions checking for primary or mapping rule access to the tenant, however it looks like we only check the mapping rule. Unless I'm over looking something we also need to consider `primaryResult.hasTenantAccess()` here too

Ben-Sheppard replied:
Hmm I'm really not a fan of that (although I can of course see that its not added by you), to me thats a hidden step that also doesn't make sense in this flow, why would I also check primary access (user/client) when I'm checking `mapping rules`? I already check their access in the first step, if they didn't have access in the first step then they won't have access in the second step, and in the opposite, if they had access in the first step then the second step I think is redundant because they already have access via their primary authentication...

I'm going to validate this with someone else on my team to check I'm aligned here and if so I'll raise an issue to clean that up but I won't block this PR on that change.

Ben-Sheppard reviewed (commented):
🚀 Thank you for addressing my comments, in general I think the changes here are good.

🔧 I have one comment that I missed the first time round and if I'm correct in my comment suggests that there is a gap in the tests. Once this is addressed I will approve and we can merge and move onto the next PR!

ce-dmelnych committed:
test: add property-based authorization tests with tenant access checks

Add tests to AuthorizationCheckBehaviorMultiTenancyTest verifying
requirement that property-based authorization only succeeds when the
user has tenant access (via primary entity or mapping rules).

Ben-Sheppard reviewed (approved):
I'm going to approve so we can make progress, we're tracking two optimizations that I think would be good to consider in time but you haven't added anything with this PR in those areas so it shouldn't block this PR :).

I think we can also raise an issue to consider the alternative that @npepinpe mentioned regarding taking the generic approach one step further but I'll raise that issue for our board (unless you already had, or planned to)

ce-dmelnych commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/authorization/AuthorizationCheckBehavior.java:
```diff
@@ -180,33 +184,41 @@ public Either<Rejection, Void> isAnyAuthorizedOrInternalCommand(
+    // Step 3: Check property-based authorization
+    // Only if we have tenant access (from primary or mapping rules)
+    if (request.hasResourceProperties() && mappingRuleResult.hasTenantAccess()) {
+      final boolean propertyAccess = checkPropertyBasedAuthorization(request, aggregatedRejections);
```
🆕 Issue for tracking: https://github.com/camunda/camunda/issues/44011
